### ISSUE TITLE
Macro: `extendsType` freezes analyzer

### ISSUE THREAD

**Issue by goderbauer** [dart-lang/sdk#55931]
Macro definition:
```dart
macro class Foo implements ClassTypesMacro {
  const Foo();

  @override
  FutureOr<void> buildTypesForClass(ClassDeclaration clazz, ClassTypeBuilder builder) async {
    final Identifier statelessWidget = await builder.resolveIdentifier(
      Uri.parse('package:flutter/src/widgets/framework.dart'),
      'StatelessWidget',
    );
    builder.extendsType(NamedTypeAnnotationCode(name: statelessWidget));
  }
}
```

Macro application:
```dart
@Foo()
class Bar { }
```

The usage of `builder.extendsType` seems to just make the analysis server get stuck. It stops reporting any issues and it also doesn't update the augmentation file anymore. Commenting out that line gets the analyzer back into a good state.

**Comment by srawlins** [comment]
CC @scheglov

**Comment by goderbauer** [comment]
Thanks for taking a look! I just tried it again on 3.5.0-222.0.dev and I am still seeing the same hanging behavior. In case it is helpful, the full code I am using is here: https://github.com/goderbauer/scratchpad/tree/main/macros_playground and when I comment in this line it just hangs: https://github.com/goderbauer/scratchpad/blob/a980fd66fdda138d72b7a04731a2537dd7f1fa47/macros_playground/lib/src/macro.dart#L157

**Comment by goderbauer** [comment]
Interesting: I tried to reproduce the hang in a fresh project with just the code given in the first comment and it doesn't repro there. But when I insert that code into my other project at https://github.com/goderbauer/scratchpad/tree/main/macros_playground it reliably hangs...

**Comment by scheglov** [comment]
I was able to reproduce it, and have a unit test that shows it.
https://dart-review.googlesource.com/c/sdk/+/371781

I believe this is a bug in the implementation that generates merged macro code.
It does not preserve the order of partial macro results.
@jakemac53

**Comment by scheglov** [comment]
This is not quite the analyzer bug, but the shared implementation issue.
So, I un-assign it from myself.
@jakemac53

**Comment by sethladd** [comment]
Thanks @scheglov for taking a look, and thank you @jakemac53 for checking it out for us!

**Comment by jakemac53** [comment]
@scheglov can you clarify why the specific ordering matters here?

I believe the intention today is we always put new types first, because that should always be valid, and any augmentations of those types must come after the original type declaration.

We could try and do something different, but it would be a very large change so I want to make sure it is actually required.

**Comment by scheglov** [comment]
I would not say that it strictly matters, maybe more like my expectations. As we run macro applications, we generate transient `*.macroX.dart` files, build `CompilationUnitElement`s for them, and at the end merge partial elements into the single `*.macro.dart` file unit element. So, my implicit expectation was that if we get partial files
```
augment library 'test.dart';

augment class B extends prefix0.A {
}
```
and then
```
class C {}
```
Then in the merged file we get declarations in the code in the same order.
And it is important to have elements and AST declaration to correspond each other in order.

Is this wrong? Your answer implies that it is wrong :-)

Theoretically one macro augmentation can both generate a new type, and augment existing, right?
So, we cannot just reorder partial library augmentations, right?
We need to merge first new types, then augmentation of classes, then functions, then augmentations of functions?
I guess I don't know if there is any better documentation of the order than the merge implementation code :)

**Comment by scheglov** [comment]
It looks to me that the best option would be for me to write multiple tests in the analyzer that cover the merging order, as well as I understand the existing implementation of `AugmentationLibraryBuilder`.

**Comment by jakemac53** [comment]
> It looks to me that the best option would be for me to write multiple tests in the analyzer that cover the merging order, as well as I understand the existing implementation of `AugmentationLibraryBuilder`.

I wouldn't focus too much on the existing behavior if possible. @johnniwinther made a lot of changes to that code to I think help with a similar issue in the CFE, attaching special identifiers to certain parts of the code? I don't fully understand the details tbh, but maybe he can help.

**Comment by scheglov** [comment]
The tests are probably still useful. Whatever order we do, the order of elements and declarations in merged code inside the analyzer must be the same. If the order changes, we will need to update the tests. How we ensure that they are consistent, is a separate question :-)

**Comment by scheglov** [comment]
OTOH, given that we don't know anymore what and how macros will generate, maybe I should just fix this specific issue and don't waste time doing more general solution.

**Comment by goderbauer** [comment]
I was able to confirm that extendsType now works flawlessly in my prototype. Thank you!

I am running into the same freezing problem at another place now and I am wondering if its related or something different. I want the macro to generate a private class (`_Foo`), but when I replace the `\$` in the line linked below with `_` everything freezes up again.

https://github.com/goderbauer/scratchpad/blob/9ea949bed426c6c9cc7e61ec15104a79bce3e659/macros_playground/lib/src/macro.dart#L17

This is what I would like that line to be (unless there's another trick to generating private classes):

```dart
String _toStatefulWidgetClassName(ClassDeclaration clazz) => '_${clazz.identifier.name}';
```

**Comment by scheglov** [comment]
There is a different crash in the analyzer, augmentation libraries are not fully implemented yet.
I opened a new issue https://github.com/dart-lang/sdk/issues/56184
